python/pyspark/rdd.py

```diff
@@ -2029,7 +2028,15 @@ def coalesce(self, numPartitions, shuffle=False):
         >>> sc.parallelize([1, 2, 3, 4, 5], 3).coalesce(1).glom().collect()
         [[1, 2, 3, 4, 5]]
         """
-        jrdd = self._jrdd.coalesce(numPartitions, shuffle)
+        if shuffle:
+            # In Scala's repartition code, we will distribute elements evenly across output
+            # partitions. However, the RDD from Python is serialized as a single binary data,
+            # so the distribution fails and produces highly skewed partitions. We need to
+            # convert it to a RDD of java object before repartitioning.
+            data_java_rdd = self._to_java_object_rdd().coalesce(numPartitions, shuffle)
```

Comment:
Hi @davies, actually it seems a simple benchmark was done in https://github.com/apache/spark/pull/15389#discussion_r82444378

If you worry, then, I'd like to proceed another benchmark with larger data and then will share when I have some time.